File: java/search-for-webmap/src/main/java/com/esri/arcgisruntime/sample/searchforwebmap/PortalItemAdapter.java
/*
 *  Copyright 2019 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.esri.arcgisruntime.sample.searchforwebmap;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.esri.arcgisruntime.portal.PortalItem;

public class PortalItemAdapter extends RecyclerView.Adapter<PortalItemAdapter.PortalItemViewHolder> {

  public interface OnItemClickListener {
    void onItemClick(PortalItem portalItem);
  }

  private final List<PortalItem> mPortalItemList;
  private final OnItemClickListener mOnItemClickListener;

  static class PortalItemViewHolder extends RecyclerView.ViewHolder {

    final TextView mPortalItemTextView;

    PortalItemViewHolder(View itemView) {
      super(itemView);
      mPortalItemTextView = itemView.findViewById(R.id.webmapItem);
    }

    void bind(final PortalItem portalItem, final OnItemClickListener listener) {
      mPortalItemTextView.setText(portalItem.getTitle());
      itemView.setOnClickListener(v -> listener.onItemClick(portalItem));
    }
  }

  public PortalItemAdapter(List<PortalItem> portalItemNames, OnItemClickListener onItemClickListener) {
    mPortalItemList = portalItemNames;
    mOnItemClickListener = onItemClickListener;
  }

  @NonNull @Override
  public PortalItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    View portalItemTextView = LayoutInflater.from(parent.getContext()).inflate(R.layout.webmap_row, parent, false);
    return new PortalItemViewHolder(portalItemTextView);
  }

  @Override
  public void onBindViewHolder(@NonNull PortalItemViewHolder holder, int position) {
    holder.bind(mPortalItemList.get(position), mOnItemClickListener);
  }

  @Override
  public int getItemCount() {
    return mPortalItemList.size();
  }
}

File: kotlin/integrated-windows-authentication/src/main/java/com/esri/arcgisruntime/sample/integratedwindowsauthentication/PortalItemAdapter.kt
/*
 * Copyright 2019 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.esri.arcgisruntime.sample.integratedwindowsauthentication

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.esri.arcgisruntime.portal.PortalItem
import com.esri.arcgisruntime.sample.integratedwindowsauthentication.databinding.PortalItemRowBinding

class PortalItemAdapter(private val onItemClickListener: OnItemClickListener) :
  RecyclerView.Adapter<PortalItemAdapter.PortalItemViewHolder>() {

  // list of PortalItems to display
  private var portalItems: MutableList<PortalItem>? = null

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PortalItemViewHolder {
      val binding = PortalItemRowBinding.inflate(LayoutInflater.from(parent.context),parent, false)
      return PortalItemViewHolder(binding)
  }

  override fun onBindViewHolder(holder: PortalItemViewHolder, position: Int) {
    holder.bind(portalItems?.get(position), onItemClickListener)
  }

  override fun getItemCount() = portalItems?.size ?: 0

  class PortalItemViewHolder(binding: PortalItemRowBinding) :
    RecyclerView.ViewHolder(binding.root) {
    private val itemTextView = binding.itemTextView

    fun bind(portalItem: PortalItem?, onItemClickListener: OnItemClickListener) {
      portalItem?.let {
        itemTextView.text = it.title
        itemView.setOnClickListener { _ ->
          onItemClickListener.onPortalItemClick(it)
        }
      }
    }
  }

  fun updatePortalItems(portalItems: List<PortalItem>) {
    if (this.portalItems == null) {
      this.portalItems = ArrayList()
    }

    DiffUtil.calculateDiff(PortalItemsDiffUtilCallback(this.portalItems, portalItems)).let {
      this.portalItems?.clear()
      this.portalItems?.addAll(portalItems)
      it.dispatchUpdatesTo(this)
    }
  }

  interface OnItemClickListener {
    fun onPortalItemClick(portalItem: PortalItem)
  }
}

class PortalItemsDiffUtilCallback(
  private val oldPortalItems: List<PortalItem>?,
  private val newPortalItems: List<PortalItem>
) : DiffUtil.Callback() {

  override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
    oldPortalItems?.get(oldItemPosition)?.itemId == newPortalItems[newItemPosition].itemId

  override fun getOldListSize() = oldPortalItems?.size ?: 0

  override fun getNewListSize() = newPortalItems.size

  override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
    oldPortalItems?.get(oldItemPosition) == newPortalItems[newItemPosition]

}
